### PR TITLE
Fix DTLS1_BAD_VER (Cisco AnyConnect VPN) support

### PR DESCRIPTION
**Pull request by dwmw2**
Various regression fixes for DTLS1_BAD_VER.

### CODE REVIEW

File: ssl/ssl_lib.c [hunk 1]
```diff
@@ -2231,7 +2231,7 @@ int SSL_export_keying_material(SSL *s, unsigned char *out, size_t olen,
                                const unsigned char *p, size_t plen,
                                int use_context)
 {
-    if (s->version < TLS1_VERSION)
+    if (s->version < TLS1_VERSION && s->version != DTLS1_BAD_VER)
```
Comment (kroeckx) [inline ssl/ssl_lib.c]: Should this check for SSL instead?

(I have no idea why that check is there.)

Comment (dwmw2) [inline ssl/ssl_lib.c]: Because for SSL3 it would end up calling `ssl_undefined_function()`, I suppose.
It's only defined for newer protocols. But that does include DTLS1_BAD_VER.

File: ssl/statem/statem_clnt.c [hunk 2]
```diff
@@ -594,6 +594,8 @@ unsigned long ossl_statem_client_max_message_size(SSL *s)
             return SERVER_HELLO_DONE_MAX_LENGTH;
 
         case TLS_ST_CR_CHANGE:
+            if (s->client_version == DTLS1_BAD_VER)
```
Comment (dwmw2) [inline ssl/statem/statem_clnt.c]: Should that be s->version ? I've never quite understood the distinction, in client code.

Comment (mattcaswell) [inline ssl/statem/statem_clnt.c]: Yes it should be. `s->version` is the negotiated version. `s->client_version` is the max version the client supports.

**Comment by mattcaswell** [comment]
Just one comment (change `client_version` to `version`) and I can plus one this.

BTW, didn't we speak about DTLS1_BAD_VER tests at some point? That would be nice :-)

**Comment by dwmw2** [comment]
Fixed. And yes, I did start working on tests at one point. Don't remember what stopped me; perhaps I was just distracted by something shiny. Or maybe I needed to reinstate/fix DTLS1_BAD_VER server support to make any of the tests work... or maybe the full handshake isn't working because we only ever *resume* sessions which are artificially created. Will take another look.

**Comment by mattcaswell** [comment]
+1

**Comment by mattcaswell** [comment]
Ping...needs another team review.

**Comment by dwmw2** [comment]
OK,  added test case. It uses the deprecated DTLSv1_client_method() in its first incarnation; subsequent commit for separate review and comment are intended to address that requirement.

**Comment by richsalz** [comment]
@mattcaswell please re-confirm and then I'll merge.

**Comment by dwmw2** [comment]
Note that I really do consider the last two commits (_"Make DTLS1_BAD_VER work with DTLS_client_method()_" and "_Honour SSL version in SSL_set_session()_") to be an RFC. I'd be a little surprised to see those merged with no discussion. Note that in RT#3711 we previously decided *not* to do a `DTLSv0_9_client_method()` on the basis that the `SSL_OP_CISCO_ANYCONNECT` hack was sufficient.... but now since that hack is no longer quite so small and self-contained I've decided that it *does* make sense to implement `DTLS1_BAD_VER` support 'properly' — albeit not with a user-visible `DTLSv0_9_client_method()` as proposed before.

But I'm still more than happy to defer to your judgement about how best to implement this. And as I noted in email earlier..,. you have a test case now :)

In general, I'd observe that there is an awful mess of ways that `s->client_version` and `s->version` are set and used, and it would be nice if that wasn't all quite so baroque.

The initial fixes, and adding the initial version of the test case that still uses the deprecated `DTLSv1_client_method()`, though, I'm happy to be merged. And one of those fixes wants merging for 1.0.2 too, as noted in RT#4631.

**Comment by dwmw2** [comment]
Test updated to actually include a recipe so it gets run with `make test`, and the `DTLS_client_method()` support fixed not to break existing DTLS tests. I hadn't properly accounted for the fact that `s->{min,max}_proto_version` might be zero.

Dropped the handling of `s->session->ssl_version` because that's a can of worms we don't want to reopen, as discussed on the mailing list.

File: test/baddtlstest.c [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+#define CIPHERS_LEN_LEN         2
+#define COMPRESSION_LEN_LEN     1
+#define EXTENSIONS_LEN_LEN      2
+#define EXTENSION_TYPE_LEN      2
+#define EXTENSION_SIZE_LEN      2
+
+
+static unsigned char client_random[SSL3_RANDOM_SIZE];
+static unsigned char server_random[SSL3_RANDOM_SIZE];
+static unsigned char session_id[32];
```
Comment (mattcaswell) [inline test/baddtlstest.c]: SSL_MAX_SSL_SESSION_ID_LENGTH

File: test/baddtlstest.c [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+static unsigned char master_secret[48];
```
Comment (mattcaswell) [inline test/baddtlstest.c]: SSL_MAX_MASTER_KEY_LENGTH

File: test/baddtlstest.c [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+    time((void *)server_random);
+
+    sess = client_session();
+    if (!sess) {
+        printf("Failed to generate SSL_SESSION\n");
+        goto end;
+    }
+
+    handshake_md = EVP_MD_CTX_new();
+    if (!handshake_md || !EVP_DigestInit_ex(handshake_md, EVP_md5_sha1(), NULL)) {
```
Comment (mattcaswell) [inline test/baddtlstest.c]: Use `handshake_md == NULL` instead of `!handshake_md`

File: test/baddtlstest.c [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+    if (!SSL_CTX_set_min_proto_version(ctx, DTLS1_BAD_VER)) {
+        printf("SSL_CTX_set_min_proto_version() failed\n");
+        goto end_ctx;
+    }
+    if (!SSL_CTX_set_max_proto_version(ctx, DTLS1_BAD_VER)) {
+        printf("SSL_CTX_set_max_proto_version() failed\n");
+        goto end_ctx;
+    }
+
+    SSL_CTX_set_read_ahead(ctx, 1);
```
Comment (mattcaswell) [inline test/baddtlstest.c]: Do you need this? `read_ahead` should not need to be set in DTLS...it should be used anyway.

File: test/baddtlstest.c [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+    unsigned char *dataend;
+    unsigned char *mac;
+    unsigned int tmplen;
+    int cookie_found = 0;
+
+    len = BIO_get_mem_data(wbio, (char **)&data);
+    dataend = data + len;
+
+    /* Check record header type, version */
+    if (data + 3 > dataend)
```
Comment (mattcaswell) [inline test/baddtlstest.c]: This idiom can actually be undefined behaviour in failure cases, i.e. where `data + 3 > dataend" is true. See:

https://www.openssl.org/blog/blog/2016/06/27/undefined-pointer-arithmetic/

Consider using ssl/packet_locl.h functions.

File: test/baddtlstest.c [hunk 8]
```diff
@@ -0,0 +1,603 @@
...
+    if (data + EXTENSIONS_LEN_LEN > dataend)
+        return 0;
+    /* Extensions len */
+    tmplen = ((*data) << 8) | *(data + 1);
+    data += EXTENSIONS_LEN_LEN;
+    if (data + tmplen != dataend)
+        return 0;
+
+    if (cookie_found)
+        if (!EVP_DigestUpdate(handshake_md, mac, dataend - mac))
```
Comment (mattcaswell) [inline test/baddtlstest.c]: Combine into one statement `cookie_found && !EVP_DigestUpdate(...)`

File: test/baddtlstest.c [hunk 9]
```diff
@@ -0,0 +1,603 @@
...
+    };
+    const unsigned char *p = session_asn1;
+
+    memcpy(session_asn1 + 15, session_id, sizeof(session_id));
+    memcpy(session_asn1 + 49, master_secret, sizeof(master_secret));
+
+    return d2i_SSL_SESSION(NULL, &p, sizeof(session_asn1));
+}
+
+static int validate_client_hello(BIO *wbio)
```
Comment (mattcaswell) [inline test/baddtlstest.c]: A comment detailing the valid return values might be helpful here.

File: test/baddtlstest.c [hunk 10]
```diff
@@ -0,0 +1,603 @@
...
+
+    /* We expect the Finished message here.*/
+    if (data + DTLS1_RT_HEADER_LENGTH > dataend)
+        return 0;
+    /* Check record header type, version */
+    if (data[0] != 0x16 || data[1] != 0x01 || data[2] != 0x00)
+        return 0;
+    /* Check epoch is now 1 */
+    if (data[3] != 00 || data[4] != 0x01)
+        return 0;
```
Comment (mattcaswell) [inline test/baddtlstest.c]: So all we're doing is checking that we've received an encrypted handshake message, with the correct version and epoch in the right place in the handshake to be an encrypted Finished? But we don't actually check that it decrypts successfully?

File: test/baddtlstest.c [hunk 11]
```diff
@@ -0,0 +1,603 @@
...
+    static unsigned char record_iv[16];
+    static unsigned char data_msg[0x20] = "Packet x";
+
+    data_hdr[10] = seq;
+    sprintf((char *)data_msg, "Packet %d", seq);
+
+    /* We want to use ssl3_cbc_digest_record() here but it's not
+     * available if we're testing against a shared library incarnation
+     * of OpenSSL. So we do it ourselves. */
+    hmac_packet(data_hdr, data_msg, 9);
```
Comment (mattcaswell) [inline test/baddtlstest.c]: So we're including the `\0`? Will fail of course if we ever change this so that `seq > 9`. Perhaps we should use `strlen` here?

File: test/baddtlstest.c [hunk 12]
```diff
@@ -0,0 +1,603 @@
...
+
+    /* Pad */
+    memset(data_msg + 29, 0x02, 3);
```
Comment (mattcaswell) [inline test/baddtlstest.c]: Magic number!

File: test/recipes/70-test_baddtls.t [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+# https://www.openssl.org/source/license.html
+
+
+use OpenSSL::Test;
+use OpenSSL::Test::Utils;
+
+setup("test_clienthello");
+
+plan skip_all => "DTLSv1 is not supported by this OpenSSL build"
+    if disabled("sctp");
```
Comment (mattcaswell) [inline test/recipes/70-test_baddtls.t]: Why SCTP? Shouldn't we be checking dtlsv1?

File: test/recipes/70-test_baddtls.t [hunk 14]
```diff
@@ -0,0 +1,20 @@
+#! /usr/bin/env perl
+# Copyright 2015-2016 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (mattcaswell) [inline test/recipes/70-test_baddtls.t]: Just 2016

File: test/baddtlstest.c [hunk 15]
```diff
@@ -0,0 +1,603 @@
+/*
+ * Copyright 2015-2016 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (mattcaswell) [inline test/baddtlstest.c]: Just 2016

**Comment by mattcaswell** [comment]
There's quite a few "magic number" which it would be nice to get rid off if possible.

**Comment by mattcaswell** [comment]
I'm feeling ok with the "Make DTLS1_BAD_VER work with DTLS_client_method()" commit.

File: test/baddtlstest.c (hunk 7 above)
Comment (dwmw2) [inline test/baddtlstest.c]: Will fix. It's inherited from the clienthello test fwiw.

File: test/baddtlstest.c (hunk 3 above)
Comment (dwmw2) [inline test/baddtlstest.c]: These are fixed in our own code as we resume a fake session. So the session ID and master secret are exactly the sizes we say.

File: test/baddtlstest.c (hunk 10 above)
Comment (dwmw2) [inline test/baddtlstest.c]: Yep. If you accepted my Finished then you are evidently getting the MAC right so far. Implementing decryption just to check you get it right for one more packet seems pointless. I don't need to check that you know how to craft a proper Finished packet; that is not a DTLS1_BAD_VER special thing so is tested elsewhere.

File: test/baddtlstest.c (hunk 11 above)
Comment (dwmw2) [inline test/baddtlstest.c]: This is not grneral-purpose code.  The packet padding is done by hand, for $DEITY's  sake. Don't use seq>9 :)

Comment (mattcaswell) [inline test/baddtlstest.c]: Ok. Perhaps just add a comment.

File: test/baddtlstest.c (hunk 10 above)
Comment (mattcaswell) [inline test/baddtlstest.c]: Ok - fair enough.

**Comment by mattcaswell** [comment]
@dwmw2 any chance you can address these issues soon? I'd like to get them in for the next beta release on Thursday.

**Comment by dwmw2** [comment]
Ack. Was away for a long weekend; dealing with them right now.

**Comment by mattcaswell** [comment]
Great! Thanks.

File: test/baddtlstest.c (hunk 6 above)
Comment (dwmw2) [inline test/baddtlstest.c]: That's still present in OpenConnect itself, working around what I think eventually became RT#3657 and got fixed in commit 8dd4ad0ff.

                /* If we don't readahead, then we do short reads and throw
                   away the tail of data packets. */
                SSL_CTX_set_read_ahead(vpninfo->dtls_ctx, 1);

Mea culpa for not filing an RT and chasing that for myself; it was [added in 2008](http://git.infradead.org/users/dwmw2/openconnect.git/commitdiff/dd4a7ff9) before I got DTLS1_BAD_VER interoperating with Cisco's protocol at all, as one of the very first _"well that's **obviously** screwed"_ things that I dealt with in my initial flailing around. I then promptly forgot about it...

File: test/baddtlstest.c [hunk 16]
```diff
@@ -0,0 +1,603 @@
...
+    if (!EVP_DigestUpdate(handshake_md, server_hello + 25, sizeof(server_hello) - 25))
+        printf("EVP_DigestUpdate() failed\n");
+
+    BIO_write(rbio, server_hello, sizeof(server_hello));
+    BIO_write(rbio, change_cipher_spec, sizeof(change_cipher_spec));
+
+    return 1;
+}
+
+static int hmac_packet(unsigned char *hdr, unsigned char *buf, int len)
```
Comment (dwmw2) [inline test/baddtlstest.c]: Oh, the irony.... after I spent three days trying to work out what the OpenSSL code was actually doing in order to reimplement the server side as a test case :)

**Comment by dwmw2** [comment]
OK, changes pushed out as an incremental patch, for ease of review _(although of course you can still just look at the final `baddtlstest.c` file if you prefer)_. I think I've captured everything, made it a little more robust even though it's only a fairly hacky test case, and left it with a *lot* fewer magic numbers.

Note that the "Make DTLS1_BAD_VER work with DTLS_client_method()" commit probably wants to go before the test case because otherwise, the test case breaks the `--strict-warnings` build due to using the deprecated `DTLSv1_client_method()`

File: test/baddtlstest.c [hunk 17]
```diff
@@ -269,70 +282,76 @@ static int send_server_hello(BIO *rbio)
...
+    HMAC_CTX *ctx;
+    EVP_CIPHER_CTX *enc_ctx;
+    unsigned char iv[16];
+    unsigned char pad = 15 - ((len + SHA_DIGEST_LENGTH) % 16);;
+    unsigned char *enc = OPENSSL_malloc(len + SHA_DIGEST_LENGTH + 1 + pad);
     unsigned char lenbytes[2];
 
-    lenbytes[0] = len >> 8;
-    lenbytes[1] = len & 0xff;
+    if (!enc)
```
Comment (mattcaswell) [inline test/baddtlstest.c]: `enc == NULL`

File: test/recipes/70-test_baddtls.t [hunk 18]
```diff
@@ -13,7 +13,7 @@ use OpenSSL::Test::Utils;
 setup("test_clienthello");
```
Comment (mattcaswell) [inline test/recipes/70-test_baddtls.t]: test_badtls

File: test/baddtlstest.c [hunk 19]
```diff
@@ -289,10 +289,12 @@ static int send_record(BIO *rbio, const unsigned char *hdr,
     HMAC_CTX *ctx;
     EVP_CIPHER_CTX *enc_ctx;
     unsigned char iv[16];
-    unsigned char pad = 15 - ((len + SHA_DIGEST_LENGTH) % 16);;
-    unsigned char *enc = OPENSSL_malloc(len + SHA_DIGEST_LENGTH + 1 + pad);
+    unsigned char pad;
+    unsigned char *enc;
     unsigned char lenbytes[2];
 
+    pad = 15 - ((len + SHA_DIGEST_LENGTH) % 16);;
```
Comment (mattcaswell) [inline test/baddtlstest.c]: double ;

File: test/baddtlstest.c [hunk 20]
```diff
@@ -0,0 +1,617 @@
...
+        return 0;
+
+    /* Now we are at the end */
+    if (PACKET_remaining(&pkt))
+        return 0;
+
+    /* Update handshake MAC for second ClientHello (with cookie) */
+    if (cookie_found && !EVP_DigestUpdate(handshake_md, data + MAC_OFFSET,
+                                          len - MAC_OFFSET))
+            printf("EVP_DigestUpdate() failed\n");;
```
Comment (mattcaswell) [inline test/baddtlstest.c]: double ;;

**Comment by mattcaswell** [comment]
3 minor issues left to fix (two instances of ";;" and a copy&paste error in the test name). Assuming those are sorted +1.

Needs another team reviewer - ping @richsalz?

**Comment by dwmw2** [comment]
Thanks for the review. I've pushed an updated version without the incremental patches, and with the 'Make DTLS1_BAD_VER work with DTLS_client_method()' commit added first. The resulting commit b05c639 is identical to the previous incremental commit 38cf6205f

**Comment by richsalz** [comment]
+1 from me, but you have to merge.

**Comment by vdukhovni** [comment]
By "merge" Rich means "rebase"...

**Comment by dwmw2** [comment]
While you're merging/rebasing, please add commit 5e37782e5ab ("Fix SSL_export_keying_material() for DTLS1_BAD_VER") to the 1.0.2 branch too. The rest aren't needed, but that one is ([RT#4631](https://rt.openssl.org/Ticket/Display.html?id=4631))

**Comment by dwmw2** [comment]
Looks like I was right to deliberately throw in an out-of-order packet RX test, despite that not being strictly a DTLS1_BAD_VER thing. It's shown up the following error with enable-ubsan:
```
../../_srcdist/test/recipes/70-test_baddtls.t .......... 
1..1
../_srcdist/ssl/record/dtls1_bitmap.c:64:9: runtime error: left shift of negative value -1
../util/shlib_wrap.sh ./baddtlstest => 1
not ok 1 - running baddtlstest
```
It goes away if I let it receive packets in-order. Will look at it shortly; I think it's a newly-discovered and pre-existing bug in OpenSSL. Yay for test cases :)

File: test/baddtlstest.c [hunk 21]
```diff
@@ -0,0 +1,617 @@
+/*
```
Comment (ekasper) [inline test/baddtlstest.c]: Completely drive-by, but please use underscores to separate words in the filename. (Not all OpenSSL C sources follow this, but a majority do.) This is `bad_dtls_test` not `badd_tls_test` :)

**Comment by dwmw2** [comment]
OK, potential fix for `satsub64be()` function added. It's late; take it with a pinch of salt...
Is this supposed to handle wrap-around of sequence numbers? It looked like it tried... but it didn't actually work, and I can't imagine it ever needing to.
I find it moderately suboptimal that `bitmap->map | 1` is almost always true, as we're wasting a bit there.

**Comment by dwmw2** [comment]
It's non-trivial to add a test case for `dtls1_record_replay_check()` directly, because it's not exported from the shared library. I could do a test which only runs in the static configuration, or even do something evil like `#include "../ssl/record/dtls1_bitmap.c"` from a test case... but actually, it's easy enough to extend my baddtlstest to do a bit more torture testing of sequence numbers. It's not quite directly bashing on `satsub64be()` like we could on a 64-bit machine, but it's not bad.

In doing so, I found some bugs in my `satsub64be()` implementation and fixed them.

**Comment by mattcaswell** [comment]
Just need to rename baddtlstest.c to bad_dtls_test.c as requested by @ekasper. Everything else looks good. Still +1 from me.

@richsalz reconfirm?

**Comment by richsalz** [comment]
+1 from me.  Matt, I suggest you merge this PR and then do the rename.  +1 from me to do that.

**Comment by dwmw2** [comment]
I already renamed in the last push. Sorry for failing to mention that explicitly.

**Comment by mattcaswell** [comment]
We missed the code freeze for the beta :-(

This will get merged after that now.

**Comment by dwmw2** [comment]
@dot-asm in commit 50ec3951dcce9b you explicitly "fixed" this code to handle wrap-arounds. Why? DTLS sequence numbers are only 48-bit anyway, aren't they? And this code isn't used from anywhere else, so it's an entirely theoretical situation...  but even if they *could* wrap around we wouldn't want them to, AFAICT.  And a quick test showed that the non-special-case code didn't actually permit a rollover from 0xFFFFFFFFFFFFFFFF to 0x0000000000000000 anyway, which looks inconsistent with the 64-bit BE case. What am I missing?

**Comment by dwmw2** [comment]
@mattcaswell I'm adding the same test case to the OpenConnect VPN client. Could I have permission to re-use bits of `packet_locl.h` (which appears to be yours) under LGPL, please? I'd rather do that than go back to my original implementation of those checks. I'll also rewrite the PRF function so it isn't copied from OpenSSL.

**Comment by mattcaswell** [comment]
You have my permission, but its not just mine. I did the original implementation but a number of people have hacked it since then. In the log I see Rich Salz (@richsalz), Emilia Kasper (@ekasper), Richard Levitte (@levitte) and Steve Henson (not on github).

**Comment by levitte** [comment]
Fine by me

**Comment by richsalz** [comment]
fine by me.

**Comment by dwmw2** [comment]
Thank you. In fact I think since I'm only using a subset it's basically just @mattcaswell and @ekasper from whom I need(ed) permission.

**Comment by ekasper** [comment]
@dwmw2, ok here too to use this file.

**Comment by dwmw2** [comment]
Great, thanks. I've [pushed](http://git.infradead.org/users/dwmw2/openconnect.git/commitdiff/18359656dfb5dce2534c95ea350a2ad8550d45e8) the test case out to OpenConnect.

And yay... CI shows test failures on RHEL where 1.0.1 and 1.0.2 are broken... :)
